lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,454 @@
+                             subBase.getFieldID() + 1 +
+                                 index *
+                                     (vecType.getElementType().getMaxFieldID() +
+                                      1)));
+                  return success();
+                },
+                false);
+            if (isValid)
+              worklist.push_back(res);
+          })
+          .Case<InstanceOp>(
+              [&](InstanceOp ins) { handleInstanceOp(ins, worklist); })
+          .Case<MemOp>([&](MemOp mem) {
+            if (!(mem.getReadLatency() == 0)) {
+              return;
+            }
+            for (auto memPort : mem.getResults()) {
+              if (!memPort.getType().isa<FIRRTLBaseType>())
                 continue;
-              storageLocIdToValuesMap[baseId + enableFieldId].push_back(
-                  dataField);
-              storageLocIdToValuesMap[baseId + addressFieldId].push_back(
-                  dataField);
+              memPorts.insert(memPort);
             }
           })
           .Default([&](auto) {});
     }
-    LLVM_DEBUG({
-      for (auto leaf : storageLocIdToValuesMap) {
-        llvm::dbgs() << "\n Valueindex :" << leaf.first;
-        for (auto val : leaf.second)
-          llvm::dbgs() << "\n leaf:" << val;
-      }
-      for (const auto &i1 : valToStorageLocIdMap) {
-        llvm::dbgs() << "\n val : " << i1.first;
-        for (const auto &i2 : i1.second) {
-          llvm::dbgs() << "\n indices :" << i2;
+    dump();
+  }
+
+  void handleInstanceOp(InstanceOp ins, SmallVector<Value> &worklist) {
+    FModuleOp refModule =
+        dyn_cast<FModuleOp>(instanceGraph.getReferencedModule(ins));
+    if (!refModule)
+      return;
+    for (auto port : ins.getResults()) {
+      worklist.push_back(port);
+      if (auto type = port.getType().dyn_cast<FIRRTLBaseType>())
+        if (!type.isGround())
+          setValRefsTo(port, Node(port, 0));
+    }
+  }
+
+  void handlePorts(Node ref, SmallVector<Value> &children) {
+    if (auto inst = dyn_cast<InstanceOp>(ref.getDefiningOp())) {
+      auto res = ref.getValue().cast<OpResult>();
+      auto portNum = res.getResultNumber();
+      auto refMod =
+          dyn_cast<FModuleOp>(instanceGraph.getReferencedModule(inst));
+      if (!refMod)
+        return;
+      Node modArg(refMod.getArgument(portNum), ref.getFieldID());
+      auto pathIter = portPaths.find(modArg);
+      if (pathIter == portPaths.end())
+        return;
+      for (auto modOutPort : pathIter->second) {
+        auto outPortNum =
+            modOutPort.getValue().cast<BlockArgument>().getArgNumber();
+        if (modOutPort.getFieldID() == 0) {
+          children.push_back(inst.getResult(outPortNum));
+          return;
         }
+        Node instanceOutPort(inst.getResult(outPortNum),
+                             modOutPort.getFieldID());
+        for (auto r : fieldToVals[instanceOutPort])
+          children.push_back(r);
```

Comment:
consider `llvm::append_range(children, fieldToVals[instanceOutPort])`, here and elsewhere similarly.